File: packages/runtime/ios/Engine/Probes.swift
import Foundation

/// The environment, behind protocols.
///
/// Detectors never touch the filesystem, `dyld` or the process environment
/// directly. They ask a probe. That one constraint is what makes every iOS
/// detector unit-testable on an ordinary macOS machine with no jailbroken device
/// anywhere in the loop — a test describes a jailbroken phone as data.
///
/// Probes are also the only place allowed to swallow a failure, and they never
/// swallow it silently: a failed read becomes `nil`, which detectors are required
/// to report as `indeterminate` rather than as "not detected".

/// Filesystem reads.
public protocol PathProbe: Sendable {
  /// Whether a path exists. `nil` when the answer cannot be determined.
  ///
  /// Implementations use `stat`/`access` rather than `FileManager`: `FileManager`
  /// is an Objective-C class and therefore a natural swizzling target, whereas
  /// the syscall wrappers are a harder — not impossible — one.
  func exists(_ path: String) -> Bool?

  /// Whether a path is a symbolic link. `nil` when undeterminable.
  func isSymbolicLink(_ path: String) -> Bool?
}

/// Sandbox boundary checks.
public protocol SandboxProbe: Sendable {
  /// Whether a write **succeeded** at a path outside the application container.
  ///
  /// The polarity matters more than anything else in this file. On a healthy
  /// sandboxed application such a write *fails*, so **success** is the jailbreak
  /// indicator. Treating failure as the indicator — as at least one widely copied
  /// implementation does — reports every healthy device as jailbroken.
  func canWriteOutsideContainer() -> Bool?
}

/// Loaded Mach-O images.
public protocol DyldProbe: Sendable {
  /// Names of images loaded into this process, or `nil` if unreadable.
  func loadedImageNames() -> [String]?
}

/// Process environment.
public protocol EnvironmentProbe: Sendable {
  /// An environment variable's value, or `nil` when unset.
  func value(for key: String) -> String?
}

/// URL scheme interrogation.
public protocol UrlSchemeProbe: Sendable {
  /// Whether the application declared `LSApplicationQueriesSchemes` at all.
  ///
  /// Without those declarations `canOpenURL` answers `false` for everything,
  /// which is a false negative indistinguishable from a clean result.
  func isConfigured() -> Bool

  /// Whether a scheme can be opened, or `nil` when undeterminable.
  func canOpen(scheme: String) -> Bool?
}

/// Process state.
public protocol ProcessProbe: Sendable {
  /// Whether this process is being traced, from `sysctl`'s `P_TRACED` flag.
  ///
  /// Public API, unlike `ptrace`, and therefore App Store safe.
  func isTraced() -> Bool?

  /// This process's parent process identifier, or `nil` if undeterminable.
  func parentProcessId() -> Int32?
}

/// Native symbol resolution.
public protocol NativeSymbolProbe: Sendable {
  /// Whether symbol resolution is available at all.
  func isAvailable() -> Bool

  /// Path of the image providing `symbol`, or `nil` if unresolvable.
  func originOf(symbol: String) -> String?
}

/// Objective-C runtime introspection.
public protocol ObjcRuntimeProbe: Sendable {
  /// Path of the image providing `selector`'s implementation on `className`.
  ///
  /// A method whose implementation lives outside the framework that declared it
  /// has been replaced by something — which is what swizzling looks like from
  /// the inside.
  func implementationImage(className: String, selector: String) -> String?
}

/// The application bundle and its main binary.
public protocol BundleProbe: Sendable {
  func bundleIdentifier() -> String?

  /// Whether an `embedded.mobileprovision` is present in the bundle.
  func hasEmbeddedProvisioningProfile() -> Bool?

  /// Whether the main binary carries App Store encryption (`cryptid`).
  func isMainBinaryEncrypted() -> Bool?
}

/// Keychain and Secure Enclave capabilities.
public protocol KeychainProbe: Sendable {
  /// Whether a Secure Enclave key can actually be created.
  ///
  /// Established by creating one and deleting it. Capability flags describe what
  /// a device *has*; only creating a key tells you what your keys will get.
  func isSecureEnclaveAvailable() -> Bool?

  /// Whether an item can be stored in and removed from the keychain.
  func canStoreKeychainItem() -> Bool?
}

/// Device authentication capabilities.
public protocol BiometryProbe: Sendable {
  /// Whether biometric authentication can be evaluated right now.
  func canEvaluateBiometrics() -> Bool?

  /// Why biometric evaluation is unavailable: `not-available`, `not-enrolled`,
  /// `lockout`, `passcode-not-set`, or `nil` when it is available.
  func biometryUnavailableReason() -> String?

  /// `none`, `touch-id`, `face-id`, `optic-id`, or `nil` if undeterminable.
  func biometryType() -> String?

  /// Whether a device passcode is set, so device-owner authentication is possible.
  func canEvaluateDeviceOwnerAuthentication() -> Bool?
}

/// Network configuration and posture.
public protocol NetworkConfigProbe: Sendable {
  /// Whether App Transport Security permits arbitrary loads for this application.
  func allowsArbitraryLoads() -> Bool?

  /// Whether a system HTTP proxy is configured.
  func isProxyConfigured() -> Bool?

  /// Whether a VPN-style network interface is present.
  func isVpnInterfacePresent() -> Bool?
}

/// Screen capture state.
public protocol ScreenCaptureProbe: Sendable {
  /// Whether the screen is currently being recorded or mirrored.
  func isCaptured() -> Bool?
}

/// Facts about the execution environment itself.
public protocol DeviceEnvironmentProbe: Sendable {
  /// Whether this process is running on the iOS Simulator.
  ///
  /// The simulator runs against the macOS filesystem, so filesystem-based
  /// jailbreak indicators are meaningless there — the question "has this iOS
  /// device been modified" is malformed when there is no iOS device.
  func isSimulator() -> Bool
}

/// Everything a detector may reach for, in one place.
public struct ProbeSet: Sendable {
  public let paths: PathProbe
  public let sandbox: SandboxProbe
  public let dyld: DyldProbe
  public let environment: EnvironmentProbe
  public let urlSchemes: UrlSchemeProbe
  public let deviceEnvironment: DeviceEnvironmentProbe
  public let process: ProcessProbe
  public let symbols: NativeSymbolProbe
  public let objcRuntime: ObjcRuntimeProbe
  public let bundle: BundleProbe
  public let keychain: KeychainProbe
  public let biometry: BiometryProbe
  public let networkConfig: NetworkConfigProbe
  public let screenCapture: ScreenCaptureProbe

  public init(
    paths: PathProbe,
    sandbox: SandboxProbe,
    dyld: DyldProbe,
    environment: EnvironmentProbe,
    urlSchemes: UrlSchemeProbe,
    deviceEnvironment: DeviceEnvironmentProbe,
    process: ProcessProbe,
    symbols: NativeSymbolProbe,
    objcRuntime: ObjcRuntimeProbe,
    bundle: BundleProbe,
    keychain: KeychainProbe,
    biometry: BiometryProbe,
    networkConfig: NetworkConfigProbe,
    screenCapture: ScreenCaptureProbe
  ) {
    self.paths = paths
    self.sandbox = sandbox
    self.dyld = dyld
    self.environment = environment
    self.urlSchemes = urlSchemes
    self.deviceEnvironment = deviceEnvironment
    self.process = process
    self.symbols = symbols
    self.objcRuntime = objcRuntime
    self.bundle = bundle
    self.keychain = keychain
    self.biometry = biometry
    self.networkConfig = networkConfig
    self.screenCapture = screenCapture
  }
}

File: packages/runtime/android/src/main/java/com/rnsecurity/probe/Probes.kt
package com.rnsecurity.probe

/**
 * The environment, behind interfaces.
 *
 * Detectors never touch the filesystem, `PackageManager` or system properties
 * directly. They ask a probe. That one constraint is what makes every detector
 * unit-testable on ordinary CI with no rooted device: a test supplies a fake
 * environment describing a rooted phone, and asserts on the signals produced.
 *
 * Probes are also the only place allowed to swallow an exception, and they never
 * swallow it silently — a failed read becomes `null`, which detectors are
 * required to report as `indeterminate` rather than as "not detected".
 */

/** Filesystem reads. */
interface FileProbe {
  /** Whether the path exists. `null` when the answer cannot be determined. */
  fun exists(path: String): Boolean?

  /** Whether the path exists and is executable. `null` when undeterminable. */
  fun isExecutable(path: String): Boolean?

  /** Reads a text file, or `null` if it cannot be read. */
  fun readText(path: String): String?

  /**
   * Attempts to create and delete a file in [directory].
   *
   * Returns `true` only if a write genuinely succeeded. `File.canWrite()` is not
   * used: it reports `false` on paths that are in fact writable on a rooted
   * device, and `true` in cases where a write would still fail.
   */
  fun canActuallyWrite(directory: String): Boolean?
}

/** Android system properties (`ro.*`, `init.svc.*`, …). */
interface SystemPropertyProbe {
  /** Whether the probe can read properties at all. */
  fun isAvailable(): Boolean

  /** Property value, or `null` if unset or unreadable. */
  fun get(key: String): String?
}

/** Installed-package queries. */
interface PackageProbe {
  /**
   * Whether package visibility has been configured for this app.
   *
   * On Android 11+ an app sees only packages declared in `<queries>`. Without
   * that declaration a package check silently returns "not installed" for
   * everything, which would be a false negative dressed up as a clean result.
   */
  fun canQueryPackages(): Boolean

  /** Whether a package is installed, or `null` if it cannot be determined. */
  fun isInstalled(packageName: String): Boolean?
}

/** `/proc` reads for the current process. */
interface ProcProbe {
  /** Contents of `/proc/self/status`, or `null` if unreadable. */
  fun selfStatus(): String?

  /** Lines of `/proc/self/maps`, or `null` if unreadable. */
  fun selfMaps(): List<String>?

  /** Lines of `/proc/self/mountinfo`, or `null` if unreadable. */
  fun selfMountInfo(): List<String>?

  /**
   * Names of this process's own threads, from `/proc/self/task/<tid>/comm`.
   *
   * Instrumentation frameworks inject worker threads with recognisable names.
   * Only this process is inspected; no other process is examined.
   */
  fun threadNames(): List<String>?
}

/** Managed-runtime introspection. */
interface RuntimeProbe {
  /** Whether a class is loadable, or `null` if the answer cannot be determined. */
  fun isClassPresent(className: String): Boolean?

  /**
   * Class names on the current call stack.
   *
   * A hooking framework that intercepts a call leaves its own frames between the
   * caller and the callee, which is visible from inside the callee.
   */
  fun currentStackClassNames(): List<String>?
}

/** Package identity and provenance. */
interface PackageIntegrityProbe {
  /**
   * SHA-256 of each signing certificate, uppercase hex, one entry per signer.
   *
   * Digested **per certificate**. Feeding several signers into one digest, as
   * some implementations do, produces an order-dependent value that cannot be
   * compared against a published fingerprint.
   */
  fun signingCertificateSha256(): List<String>?

  /** Package that installed this application, or `null` if undeterminable. */
  fun installerPackageName(): String?

  /** Filesystem location of the running APK, or `null` if undeterminable. */
  fun applicationSourceDir(): String?
}

/** Android Keystore capabilities. */
interface KeystoreProbe {
  /**
   * Security level backing a freshly generated key.
   *
   * One of `software`, `trusted-environment`, `strongbox`, or `unknown`.
   * `null` when the keystore could not be exercised at all.
   *
   * Determined by generating a throwaway key, inspecting it, and deleting it —
   * capability flags alone do not tell you what a key you actually create will
   * be backed by.
   */
  fun keySecurityLevel(): String?
}

/** Device authentication capabilities. */
interface AuthenticationProbe {
  /**
   * Raw `BiometricManager` status for any enrolled biometric, or `null` when
   * unavailable on this API level. Platform codes are returned unchanged so the
   * detector does the interpreting, where it can be tested.
   */
  fun biometricStatus(): Int?

  /** Raw `BiometricManager` status restricted to Class 3 (strong) biometrics. */
  fun strongBiometricStatus(): Int?

  /** Whether a PIN, pattern or password is set, or `null` if undeterminable. */
  fun isDeviceSecure(): Boolean?

  /**
   * Why a biometric status is unavailable, or `null` when one was obtained.
   *
   * One of `api-level`, `permission`, or `service-unavailable`. Without this an
   * application sees "could not be determined" and has nothing to act on — and
   * the most common cause, a missing `USE_BIOMETRIC` declaration, is entirely
   * within the application's control.
   */
  fun biometricUnavailableReason(): String?
}

/** Network configuration and posture. */
interface NetworkProbe {
  /** Whether this application is permitted to send cleartext HTTP. */
  fun isCleartextTrafficPermitted(): Boolean?

  /** Configured HTTP proxy host, or `null` when none is set. */
  fun httpProxyHost(): String?

  /**
   * Whether a VPN transport is active.
   *
   * `null` when it cannot be determined — most often because the application has
   * not declared `ACCESS_NETWORK_STATE`.
   */
  fun isVpnActive(): Boolean?

  /** Number of user-added certificate authorities in the system trust store. */
  fun userAddedCaCount(): Int?
}

/** Screen capture protection state. */
interface ScreenProbe {
  /**
   * Whether `FLAG_SECURE` is set on the current window.
   *
   * `null` when there is no current activity to inspect — during a cold start,
   * or while the application is in the background.
   */
  fun isSecureFlagSet(): Boolean?
}

/** Native symbol resolution. */
interface NativeSymbolProbe {
  /** Whether native symbol resolution is available at all. */
  fun isAvailable(): Boolean

  /** Path of the shared library providing [symbol], or `null` if unresolvable. */
  fun originOf(symbol: String): String?
}

/** Build fingerprint values. Separated so tests can describe any device. */
interface BuildProbe {
  fun tags(): String?

  fun fingerprint(): String?

  fun model(): String?

  fun manufacturer(): String?

  fun brand(): String?

  fun product(): String?

  fun device(): String?

  fun hardware(): String?

  fun board(): String?

  fun sdkInt(): Int
}

/** Debugger attachment state. */
interface DebuggerProbe {
  /** Whether a debugger is currently attached, or `null` if undeterminable. */
  fun isDebuggerConnected(): Boolean?

  /** Whether the process is blocked waiting for a debugger to attach. */
  fun isWaitingForDebugger(): Boolean?
}

/** Facts about the host application itself. */
interface ApplicationProbe {
  /** Whether the application is built debuggable, or `null` if undeterminable. */
  fun isDebuggable(): Boolean?

  /** The application's own package name. */
  fun packageName(): String?
}

/** Device-wide settings (`Settings.Global`). */
interface SettingsProbe {
  /**
   * Reads an integer from `Settings.Global`.
   *
   * Returns [defaultValue] when the key is absent, because on Android an absent
   * setting means "the platform default applies" rather than "unknown" — most
   * devices never write these keys at all, and reporting that as inconclusive
   * would make the check useless.
   *
   * Returns `null` only when the settings provider itself could not be read,
   * which genuinely is inconclusive.
   */
  fun globalInt(key: String, defaultValue: Int): Int?
}

/** Hardware and platform capabilities. */
interface DeviceFeatureProbe {
  /** Whether a `PackageManager.FEATURE_*` is present, or `null` if undeterminable. */
  fun hasSystemFeature(feature: String): Boolean?

  /** `TelephonyManager.PHONE_TYPE_*`, or `null` if undeterminable. */
  fun phoneType(): Int?

  /** Number of sensors the platform reports, or `null` if undeterminable. */
  fun sensorCount(): Int?
}

/** Everything a detector may reach for, in one place. */
data class ProbeSet(
  val files: FileProbe,
  val properties: SystemPropertyProbe,
  val packages: PackageProbe,
  val proc: ProcProbe,
  val build: BuildProbe,
  val debugger: DebuggerProbe,
  val application: ApplicationProbe,
  val device: DeviceFeatureProbe,
  val settings: SettingsProbe,
  val runtime: RuntimeProbe,
  val symbols: NativeSymbolProbe,
  val packageIntegrity: PackageIntegrityProbe,
  val keystore: KeystoreProbe,
  val authentication: AuthenticationProbe,
  val network: NetworkProbe,
  val screen: ScreenProbe
)
